apis/v1alpha1/gateway_types.go

```diff
@@ -483,47 +485,125 @@ type GatewayStatus struct {
+	// * "Ready"
+	//
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=type
+	// +kubebuilder:default={{type: "Scheduled", status: "False", reason:"NotReconciled", message:"Waiting for controller", lastTransitionTime: "1970-01-01T00:00:00Z"}}
 	Conditions []metav1.Condition `json:"conditions,omitempty"`
 
 	// Listeners provide status for each unique listener port defined in the Spec.
+	//
 	// +optional
 	Listeners []ListenerStatus `json:"listeners,omitempty"`
 }
 
 // GatewayConditionType is a type of condition associated with a
 // Gateway. This type should be used with the GatewayStatus.Conditions
 // field.
+//
 type GatewayConditionType string
 
-const (
-	// ConditionForbiddenNamespaceForClass indicates that this Gateway is in
-	// a namespace forbidden by the GatewayClass.
-	ConditionForbiddenNamespaceForClass GatewayConditionType = "ForbiddenNamespaceForClass"
+// GatewayConditionReason defines the set of reasons that explain
+// why a particular Gateway condition type has been raised.
+type GatewayConditionReason string
 
+const (
 	// ConditionGatewayNotScheduled indicates that the Gateway has not been
 	// scheduled.
-	ConditionGatewayNotScheduled GatewayConditionType = "GatewayNotScheduled"
+	ConditionGatewayNotScheduled GatewayConditionType = "NotScheduled"
 
 	// ConditionListenersNotReady indicates that at least one of the specified
 	// listeners is not ready. If this condition has a status of True, a more
 	// detailed ListenerCondition should be present in the corresponding
 	// ListenerStatus.
 	ConditionListenersNotReady GatewayConditionType = "ListenersNotReady"
+)
 
-	// ConditionInvalidListeners indicates that at least one of the specified
-	// listeners is invalid. If this condition has a status of True, a more
-	// detailed ListenerCondition should be present in the corresponding
-	// ListenerStatus.
-	ConditionInvalidListeners GatewayConditionType = "InvalidListeners"
+const (
+	// GatewayConditionScheduled indicates whether the controller
+	// managing the Gateway has scheduled the Gateway to the
+	// underlying network infrastructure.
+	//
+	// Valid reasons for this condition to be false are:
+	//
+	// * "NotReconciled"
+	// * "NamespaceForbidden"
+	// * "NoSuchGatewayClass"
+	// * "NoResources"
+	GatewayConditionScheduled GatewayConditionType = "Scheduled"
+
+	// GatewayReasonNotReconciled is used when the Gateway is
+	// not scheduled because it recently been created and no
+	// controller has reconciled it yet.
+	GatewayReasonNotReconciled GatewayConditionReason = "NotReconciled"
+
+	// GatewayReasonNamespaceForbidden is used when the Gateway is
+	// not scheduled because, while the Gateway is visible to
+	// the managing controller, the controller's configuration
+	// forbids is from scheduling objects from that namespace.
+	GatewayReasonNamespaceForbidden GatewayConditionReason = "NamespaceForbidden"
+
+	// GatewayReasonNoSuchGatewayClass is used when the Gateway is
```

Comment:
What happens if multiple controllers are trying to write to this condition? Any chance we'd end up with some kind of race here?

Reply:
This isn't special. If multiple controllers want to write this condition, they will race and interfere with each other.

Reply:
That's kinda what I'd been thinking as well. Although it seems unlikely, it seems like one poorly behaving controller here could cause any other good controllers to spend an awful lot of time updating this condition. Unless we can find a way to set this condition safely I'd be tempted to remove it for now.

Reply:
I don't think that I see the problem. It's already possible for badly behaved controllers to whack each others fields in bad ways. A well-behaved controller would never set this reason on a condition or even touch a gateway that didn't match it class. I think that the service-apis project should provide a controller that sets this.

Reply:
That's an interesting idea, do you think there are other conditions a common/shared controller would be responsible for setting? I'm hesitant to add one more component to maintain here, but if it seems like it might be worth it if it helped avoid potential race conditions like this.

Reply:
> That's an interesting idea, do you think there are other conditions a common/shared controller would be responsible for setting?

I think the other thing that has been discussed is a validating admission controller.

>  I'm hesitant to add one more component to maintain here, but if it seems like it might be worth it if it helped avoid potential race conditions like this.

I'm not yet convinced there is a real race here. This reason is intended for a dedicated controller, since actual ingress controllers should only even be touching gateways of their own class. I updated the description to add some more discouragement.